jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BhandariKDisjointShortestPaths.java

```diff
@@ -0,0 +1,316 @@
+            if (cPath > 1) {
+                prepare(this.pathList.get(cPath - 2));
+            }                       
+            bellmanFordShortestPath = new BellmanFordShortestPath<>(workingGraph);
+            currentPath = bellmanFordShortestPath.getPath(startVertex, endVertex);
+            if (currentPath != null) {
+                pathList.add(currentPath.getEdgeList());
+            } else {
+                break;
+            }
+        }
+
+        return pathList.size() > 0 ? resolvePaths(startVertex, endVertex) : Collections.emptyList();
+    }
+    
+    /**
+     * Prepares the graph for a search of the next path:
+     * Replacing the edges of the previous path with reversed edges
+     * with negative weight
+     * 
+     * @param cPath the number of the next path to search 
+     */
+    private void prepare(List<E> previousPath) {
+        
+        V source, target;
+        E reversedEdge;
+        
+        //replace previous path edges with reversed edges with negative weight
+        for (E originalEdge : previousPath) {
+            source = workingGraph.getEdgeSource(originalEdge);
+            target = workingGraph.getEdgeTarget(originalEdge);
+            workingGraph.removeEdge(originalEdge);    
+            reversedEdge = workingGraph.addEdge(target, source);
+            if (reversedEdge != null) {
+                workingGraph.setEdgeWeight(reversedEdge, - workingGraph.getEdgeWeight(originalEdge));
+            }
+        }
+    }
+    
+    /**
+     * At the end of the search we have list of intermediate paths - not necessarily
+     * disjoint and may contain reversed edges. Here we go over all, removing overlapping
+     * edges and merging them to valid paths (from start to end). Finally, we sort
+     * them according to their weight.
+     * 
+     * @param endVertex the end vertex
+     * 
+     * @return sorted list of disjoint paths from start vertex to end vertex.
+     */
+    private List<GraphPath<V, E>> resolvePaths(V startVertex, V endVertex) {
+        //first we need to remove overlapping edges.        
+        findOverlappingEdges();
+        
+        //now we might be left with path fragments (not necessarily leading from start to end).
+        //We need to merge them to valid paths.
+        List<GraphPath<V, E>> paths = mergePaths(startVertex, endVertex);
+        
+        //sort paths by overall weight (ascending)
+        Collections.sort(paths, (o1, o2) -> Double.compare(o1.getWeight(), o2.getWeight()));        
+        return paths;
+    }
+    
+    /**
+     * After removing overlapping edges, each path is not necessarily connecting
+     * start to end vertex. Here we connect the path fragments to valid paths
+     * (from start to end).
+     * 
+     * @param endVertex the end vertex
+     * 
+     * @return list of disjoint paths from start to end.
+     */    
+    private List<GraphPath<V, E>> mergePaths(V startVertex, V endVertex)

```

Comment:
To my understanding, all edges (excluding the 'overLappingEdges') form exactly k-disjoint paths. In your code, it seems you try to iteratively try to append an edge to any of the k paths, until all edges are used in some path. This seems quite inefficient, since you already know that the paths are disjoint. Consider the following (untested) code:
```
public void buildPaths(){
    List<E> nonOverlappingEdges=...; //see comment below
    List<List<E>> paths=new ArrayList<>();
    Map<V,Queue<E>> sourceToEdgeLookup=new HashMap<>();
    for(E e : nonOverlappingEdges){
        V u=workingGraph.getEdgeSource(e);
        if(u.equals(startVertex)) //start of a new path
            paths.add(Arrays.asList(e));
        else { //some edge which is part of a path
            if (!sourceToEdgeLookup.containsKey(u))
                sourceToEdgeLookup.put(u, new ArrayDeque<>());
            sourceToEdgeLookup.get(u).add(e);
        }
    }
    //Build the paths using the lookup table
    for(List<E> path : paths){
        V v =workingGraph.getEdgeTarget(path.get(0));
        while(!v.equals(endVertex)){
            E e=sourceToEdgeLookup.get(v).poll();
            path.add(e);
            v=workingGraph.getEdgeTarget(e);
        }
    }
    //Convert the paths to graphWalks
    ...
}
```
Notice that this method also makes your `flatPathListOrdered` method obsolete, thereby saving a lot of computations.

Reply:
Ok. Two questions: 
1. This has better running time but consumes more space due to `sourceToEdgeLookup` map, correct me if I'm wrong. So, we would usually prefer time over space? 
2. We don't really need `sourceToEdgeLookup` value type to be `Queue`. We can use simple `List`, don't we? I previously needed the queue only to `peekLast`.

Reply:
1. All edge objects are already present in memory. So your map will only contain references to objects you already had. Consequently, your space increase will be incredibly small, but your speed increase can be very significant...
2. Better go for a Stack for efficiency. You don't need a list you can index.